iminuit/_libiminuit.pyx

```diff
@@ -24,10 +24,207 @@ np.import_array()
 
 __all__ = ['Minuit']
 
-# Our wrappers
+# Pointer types
 ctypedef FCNGradientBase* FCNGradientBasePtr
 ctypedef IMinuitMixin* IMinuitMixinPtr
 ctypedef PythonGradientFCN* PythonGradientFCNPtr
+ctypedef MnUserParameterState* MnUserParameterStatePtr
+
+# Helper functions
+cdef setParameterState(MnUserParameterStatePtr state, object parameters, dict fitarg):
+    """Construct parameter state from user input.
+
+    Caller is responsible for cleaning up the pointer.
+    """
+    cdef double inf = float("infinity")
+    cdef double val
+    cdef double err
+    cdef double lb
+    cdef double ub
+    for i, pname in enumerate(parameters):
+        val = fitarg[pname]
+        err = fitarg['error_' + pname]
+        state.Add(pname, val, err)
+
+        lim = fitarg['limit_' + pname]
+        if lim is not None:
+            lb = -inf if lim[0] is None else lim[0]
+            ub = inf if lim[1] is None else lim[1]
+            if lb >= ub:
+                raise ValueError(
+                    'limit for parameter %s is invalid. %r' % (pname, (lb, ub)))
+            if lb == -inf and ub == inf:
+                pass
+            elif ub == inf:
+                state.SetLowerLimit(i, lb)
+            elif lb == -inf:
+                state.SetUpperLimit(i, ub)
+            else:
+                state.SetLimits(i, lb, ub)
+            #need to set value again
+            #taking care of internal/external transformation
+            state.SetValue(i, val)
+            state.SetError(i, err)
+
+        if fitarg['fix_' + pname]:
+            state.Fix(i)
+
+
+cdef auto_frontend():
```

Comment:
A question about this `auto_frontend`: why do we set HTML frontend like this? Won't this be set if we're in `ipython` terminal where no HTML output should be used? Shouldn't the check be rather for Jupyter somehow? (If needed at all, I think they have a special method based interface instead to emit rich repr content, maybe we should use that?

It's not a change you made here, just a question because I see this code now appear in the diff.

Reply:
Not a change I made, but it is something I would like to see fixed as well. Opened an issue about it.